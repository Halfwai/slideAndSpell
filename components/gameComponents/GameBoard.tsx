import React, { useState, useEffect, useRef } from "react";
import { StyleSheet, Dimensions, Animated, View } from "react-native";

// Import gameBoardFunctions
import { checkSlidable, checkWords, checkWord, findZero, getFinalWord, removeZero, returnZeroPos, switchZero } from "@/utils/gameBoardFunctions";

// Import components
import GameBoardTile from "@/components/gameComponents/GameBoardTile";
import EmptySquare from "@/components/gameComponents/EmptySquare";
import ExtraTile from "@/components/gameComponents/ExtraTile";
import ScoreComponent from "@/components/submenuComponents/ScoreComponent";
import Definitions from "@/components/submenuComponents/Definitions";
import Hints from "@/components/gameComponents/Hints";

// Setup props
interface GameBoardProps {
    gameBoard: string[][],
    extraLetter: string,
    onGameEnd: Function,
    hints?: string[][],
    returnCompleteWords?: Function
}

// Define the Gameboard component
export default function GameBoard(props: GameBoardProps) {
    // Calculate the size of the board and tiles
    const size = props.gameBoard.length;
    const spaceSize = ((Dimensions.get("window").width * 0.85) - 2) / size;
    const tileSize = spaceSize - 4;

    // Setup the state variables
    const [squareColour, setSquareColour] = useState(0);
    const boardPosition = useRef(new Animated.Value(Dimensions.get('window').height)).current;
    const [slides, setSlides] = useState(0);
    const [incrementTime, setIncrementTime] = useState(false);
    const [validWords, setValidWords] = useState<{ word: string, definition: string }[]>([]);
    const [activateSlideUp, setActivateSlideUp] = useState(false);
    const [board, setBoard] = useState(props.gameBoard);
    const [validArray, setValidArray] = useState(Array(size).fill(false));
    const [zeroPos, setZeroPos] = useState(returnZeroPos(board, spaceSize));
    const [canInsertLetter, setCanInsertLetter] = useState(false);
    const [gameOver, setGameOver] = useState(false);
    const [hints, setHints] = useState<string[][]>([]);
    const [hintsSlideIn, setHintsSlideIn] = useState(false);

    // Slide the board in
    useEffect(() => {
        Animated.timing(boardPosition, {
            toValue: 0,
            duration: 1000,
            useNativeDriver: true
        }).start(({ finished }: { finished: boolean }) => {
            if (finished) setIncrementTime(true);
        });
    }, []);

    // Function to slide the board out
    function slideBoardOut() {
        Animated.timing(boardPosition, {
            toValue: Dimensions.get('window').height,
            duration: 500,
            useNativeDriver: true
        }).start(({ finished }: { finished: boolean }) => {
            if (finished) setActivateSlideUp(true);
        });
    }

    // Run the checkBoard function when the board state changes
    useEffect(() => {
        checkBoard();
    }, [board]);

    // Function to check the board
    const checkBoard = () => {
        // Find the zero position and check the words
        setZeroPos(returnZeroPos(board, spaceSize));
        const { correctWords, newValidArray } = checkWords(board);
        // Return the number of complete words
        props.returnCompleteWords && props.returnCompleteWords(correctWords.length);
        // If the validArray has changed, update the state
        if (validArray !== newValidArray) {
            setValidArray(newValidArray);
        }
        // If the correctWords array is full, check the final word
        if (correctWords.length === size - 1 && !gameOver) {
            const { word, zeroY } = getFinalWord(board, props.extraLetter);
            const definition = checkWord(word);
            if (definition) {
                correctWords.splice(zeroY, 0, { word: word, definition: definition });
                setValidWords(correctWords);
                setCanInsertLetter(true);
            } else {
                setCanInsertLetter(false);
            }
        }
        // Adds hints to the board depending on the number of slides
        if (props.hints && hints.length < props.hints.length && slides % 10 === 0 && slides !== 0) {
            setHints((hints) => {
                const newHints = [...hints];
                if (props.hints) {
                    newHints.push(props.hints[hints.length]);
                }
                return newHints;
            });
            setHintsSlideIn(true);
        }
    }

    return (
        <>
            <Animated.View style={[styles().mainBoard, { transform: [{ translateY: boardPosition }] }]}>
                <Hints
                    startSlideIn={hintsSlideIn}
                    hints={hints}
                    tileSize={tileSize}
                    spaceSize={spaceSize}
                />
                {/* Map the board state to Tile components */}
                {board.map((row, i) => (
                    row.map((cell, j) => (
                        <React.Fragment
                            key={`${i}:${j}`}
                        >
                            {findZero(board).x === j && findZero(board).y === i ?
                                <EmptySquare
                                    tileSize={tileSize}
                                    position={{ x: j * spaceSize, y: i * spaceSize }}
                                    colour={squareColour}
                                /> :
                                <GameBoardTile
                                    value={cell}
                                    position={{ x: j * spaceSize, y: i * spaceSize }}
                                    spaceSize={spaceSize}
                                    tileSize={tileSize}
                                    slidable={checkSlidable(i, j, board)}
                                    switch={() => {
                                        setSlides((slides) => slides + 1);
                                        return switchZero(i, j, board);
                                    }}
                                    resetBoard={(newBoard: string[][]) => {
                                        setBoard(newBoard);
                                    }}
                                    valid={validArray[i]}
                                    disabled={props.hints === undefined && j < size - 1}
                                />
                            }
                        </React.Fragment>
                    ))
                ))}
            </Animated.View>
            {!gameOver && props.extraLetter && <ExtraTile
                letter={props.extraLetter}
                tileSize={tileSize}
                spaceSize={spaceSize}
                boardSize={size}
                zeroPos={zeroPos}
                canInsert={canInsertLetter}
                setEmptySquareColour={(colour: number) => {
                    setSquareColour(colour)
                }}
                removeZero={() => {
                    setBoard(removeZero(board, props.extraLetter));
                    setGameOver(true);
                    setIncrementTime(false);
                    setTimeout(() => {
                        slideBoardOut();
                    }, 1000);
                }}
            />}
            <ScoreComponent
                incrementTime={incrementTime}
                slides={slides}
                gameOver={gameOver}
                returnTime={(time: number) => {
                    props.onGameEnd(time, slides, board);
                }}
                slideUp={activateSlideUp}
            />
            {gameOver && validWords.length > 0 &&
                <View style={{ alignItems: 'center', justifyContent: 'center', width: Dimensions.get('screen').width, height: Dimensions.get('screen').height * 0.6 }}>
                    <Definitions
                        slideIn={activateSlideUp}
                        validWords={validWords}
                    />
                </View>
            }
        </>
    );
}

// Define the styles
const styles = () => StyleSheet.create({
    mainBoard: {
        width: (Dimensions.get("screen").width * 0.85),
        height: (Dimensions.get("screen").width * 0.85),
        flexDirection: 'row',
        borderWidth: 2,
        borderRadius: 10,
        backgroundColor: '#FFFFFF',
        zIndex: 1,
        position: 'absolute',
    }
});