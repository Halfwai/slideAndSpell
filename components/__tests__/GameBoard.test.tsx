import React from 'react';
import { render } from '@testing-library/react-native';

import GameBoard from '@/components/gameComponents/GameBoard';

it(`GameBoard Component renders correctly`, () => {
    const { toJSON } = render(
        <GameBoard 
            gameBoard={[
                ["A", "B", "C", "D"],
                ["E", "F", "G", "H"],
                ["I", "J", "K", "L"],
                ["M", "N", "O", "0"]
            ]}
            extraLetter={"P"}
            onGameEnd={() => {}}
            hints={[
                ["A", "B", "C", "D"],
                ["E", "F", "G", "H"],
                ["I", "J", "K", "L"],
                ["M", "N", "O", "P"]
            ]}
            returnCompleteWords={() => {}}
        />
    );
    expect(toJSON()).toMatchSnapshot();
});