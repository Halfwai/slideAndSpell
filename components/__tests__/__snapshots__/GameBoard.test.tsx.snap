// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard Component renders correctly 1`] = `
[
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 637.5,
        "position": "absolute",
        "transform": [
          {
            "translateY": 1334,
          },
        ],
        "width": 637.5,
        "zIndex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "position": "absolute",
          "transform": [
            {
              "translateX": -750,
            },
            {
              "translateY": -150,
            },
          ],
          "zIndex": 1,
        }
      }
    />
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 0,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          A
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 158.875,
            },
            {
              "translateY": 0,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          B
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 317.75,
            },
            {
              "translateY": 0,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          C
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 476.625,
            },
            {
              "translateY": 0,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          D
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 158.875,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          E
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 158.875,
            },
            {
              "translateY": 158.875,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          F
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 317.75,
            },
            {
              "translateY": 158.875,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          G
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 476.625,
            },
            {
              "translateY": 158.875,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          H
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 317.75,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          I
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 158.875,
            },
            {
              "translateY": 317.75,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          J
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 317.75,
            },
            {
              "translateY": 317.75,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          K
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 476.625,
            },
            {
              "translateY": 317.75,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          L
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 476.625,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          M
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 158.875,
            },
            {
              "translateY": 476.625,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          N
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "backgroundColor": "rgba(170, 170, 170, 1)",
          "borderRadius": 10,
          "height": 154.875,
          "margin": 1,
          "position": "absolute",
          "transform": [
            {
              "translateX": 317.75,
            },
            {
              "translateY": 476.625,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          O
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 10,
          "height": 154.875,
          "transform": [
            {
              "translateX": 476.625,
            },
            {
              "translateY": 476.625,
            },
          ],
          "width": 154.875,
          "zIndex": -1,
        }
      }
    />
  </View>,
  <View
    style={
      {
        "alignItems": "center",
        "position": "absolute",
        "top": -476.625,
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <View
      collapsable={false}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#AAA",
          "borderRadius": 10,
          "height": 154.875,
          "justifyContent": "center",
          "margin": 1,
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 0,
            },
          ],
          "width": 154.875,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontFamily": "Postino_std",
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          P
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "flexDirection": "row",
        "marginTop": 20,
        "position": "absolute",
        "zIndex": 0,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#8BF0BC",
          "borderBottomWidth": 5,
          "borderColor": "white",
          "borderTopWidth": 5,
          "flexDirection": "column",
          "justifyContent": "center",
          "paddingVertical": 10,
          "transform": [
            {
              "translateX": -750,
            },
            {
              "translateY": 200,
            },
          ],
          "width": 375,
        }
      }
    >
      <Text
        style={
          {
            "color": "black",
            "fontSize": 20,
            "textAlign": "center",
          }
        }
      >
        Time
      </Text>
      <Text
        style={
          {
            "color": "black",
            "fontSize": 20,
            "textAlign": "center",
          }
        }
      >
        00:00:00
      </Text>
    </View>
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#8BF0BC",
          "borderBottomWidth": 5,
          "borderColor": "white",
          "borderTopWidth": 5,
          "flexDirection": "column",
          "justifyContent": "center",
          "paddingVertical": 10,
          "transform": [
            {
              "translateX": 750,
            },
            {
              "translateY": 200,
            },
          ],
          "width": 375,
        }
      }
    >
      <Text
        style={
          {
            "color": "black",
            "fontSize": 20,
            "textAlign": "center",
          }
        }
      >
        Slides
      </Text>
      <Text
        style={
          {
            "color": "black",
            "fontSize": 20,
            "textAlign": "center",
          }
        }
      >
        0
      </Text>
    </View>
  </View>,
]
`;
